## Data Analysis I

Lastly, we will answer the following questions using pandas methods: 

* What are the most popular colors by season?
* What is the most popular clothing item by season?
* What is the effect of promo-codes on the dollar-amount of purchases?
* When do users leave a review?
* Do states vary in the items they purchase in the Fall?

Utilize the documentation provided in each code-block. When you are done with this section of the project, validate that your output matches the screenshot provided in the `docs/part3.md` file and answer the questions located underneath `Data Analysis II` in your own words.

In [16]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

In [17]:
# Load `data/processed/shopping_cleaned.csv` as a pandas dataframe

df = pd.read_csv('../data/processed/shopping_cleaned.csv')
df

Age  Gender Item Purchased  Purchase Amount (USD)      Location Size  \
0     31.0    Male       Leggings              24.231704        Nevada    M   
1     38.0    Male     Sunglasses              36.545487        Oregon    S   
2     26.0    Male       Leggings              23.595139        Nevada   XL   
3     43.0    Male          Dress              34.079015    California    M   
4     29.0    Male         Shorts              23.796439     Minnesota    M   
...    ...     ...            ...                    ...           ...  ...   
3153  39.0    Male  Running Shoes              55.636680        Kansas    L   
3154  43.0  Female     Sunglasses              61.610602      Colorado    S   
3155  37.0    Male        Pajamas              44.600556        Alaska    S   
3156  39.0    Male         Hoodie              45.343778      Illinois    S   
3157  21.0  Female         Gloves              49.439181  North Dakota    M   

             Color  Season Review Rating   Shipping Type Promo Code Used  \
0      Terra cotta  Winter       Present        Standard              No   
1        Aubergine  Summer       Missing        Standard              No   
2            Brown  Winter       Missing        Standard              No   
3      Terra cotta    Fall       Present        Standard              No   
4         Lavender  Summer       Present         Express              No   
...            ...     ...           ...             ...             ...   
3153         White  Summer       Missing  2-Day Shipping             Yes   
3154  Burnt orange    Fall       Missing  2-Day Shipping             Yes   
3155     Aubergine  Winter       Present  2-Day Shipping             Yes   
3156     Brick red    Fall       Missing  2-Day Shipping             Yes   
3157     Aubergine  Winter       Present  2-Day Shipping             Yes   

      Previous Purchases Payment Method  
0                      0    Credit Card  
1                      0    Credit Card  
2                      3    Credit Card  
3                      0    Credit Card  
4                      0    Credit Card  
...                  ...            ...  
3153                  11          Venmo  
3154                   7          Venmo  
3155                  20          Venmo  
3156                   3          Venmo  
3157                   2          Venmo  

[3158 rows x 13 columns]

In [18]:
# Create a dataframe pivot table where "Season" is your column, "Color" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

pivot1 = pd.pivot_table(data = df, columns = 'Color', index = 'Season', values = 'Location', aggfunc= len)


In [19]:
# Display this pivot table

pivot1

Color   Aubergine  Baby blue  Black  Brick red  Brown  Burnt orange  Fuchsia  \
Season                                                                         
Fall           83         16     84         88     97            93        3   
Spring          2         69      5          7      5             4       45   
Summer         13         84      9         12      3             5       78   
Winter        111         21    124        110     89           112       18   

Color   Lavender  Lemon yellow  Mauve  Muted mustard yellow  Pale peach  \
Season                                                                    
Fall          13             9     86                    84           9   
Spring        55            47      5                     8          52   
Summer       102            98      7                    13          81   
Winter         6            13     95                    87          15   

Color   Periwinkle  Ruby Red  Terra cotta  Turquoise  White  
Season                                                       
Fall             9        90           91          7     10  
Spring          54         3            4         49     63  
Summer          82         8           15         74     81  
Winter           8       107          101         16     11

In [20]:
# Select the "Fall" column from this pivot table and display the sort order in descending order
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

pivot1.sort_values(['Fall'],ascending= False)

KeyError: 'Fall'

In [ ]:
# Select the "Winter" column from this pivot table and display the sort order in descending order

pivot1.sort_values(['Winter'], ascending = False)

Season                Fall  Spring  Summer  Winter
Color                                             
Black                   84       5       9     124
Burnt orange            93       4       5     112
Aubergine               83       2      13     111
Brick red               88       7      12     110
Ruby Red                90       3       8     107
Terra cotta             91       4      15     101
Mauve                   86       5       7      95
Brown                   97       5       3      89
Muted mustard yellow    84       8      13      87
Baby blue               16      69      84      21
Fuchsia                  3      45      78      18
Turquoise                7      49      74      16
Pale peach               9      52      81      15
Lemon yellow             9      47      98      13
White                   10      63      81      11
Periwinkle               9      54      82       8
Lavender                13      55     102       6

In [ ]:
# Select the "Spring" column from this pivot table and display the sort order in descending order

pivot1.sort_values(['Spring'], ascending= False)

Season                Fall  Spring  Summer  Winter
Color                                             
Baby blue               16      69      84      21
White                   10      63      81      11
Lavender                13      55     102       6
Periwinkle               9      54      82       8
Pale peach               9      52      81      15
Turquoise                7      49      74      16
Lemon yellow             9      47      98      13
Fuchsia                  3      45      78      18
Muted mustard yellow    84       8      13      87
Brick red               88       7      12     110
Mauve                   86       5       7      95
Brown                   97       5       3      89
Black                   84       5       9     124
Burnt orange            93       4       5     112
Terra cotta             91       4      15     101
Ruby Red                90       3       8     107
Aubergine               83       2      13     111

In [ ]:
# Select the "Summer" column from this pivot table and display the sort order in descending order

pivot1.sort_values(['Summer'], ascending = False)

Season                Fall  Spring  Summer  Winter
Color                                             
Lavender                13      55     102       6
Lemon yellow             9      47      98      13
Baby blue               16      69      84      21
Periwinkle               9      54      82       8
Pale peach               9      52      81      15
White                   10      63      81      11
Fuchsia                  3      45      78      18
Turquoise                7      49      74      16
Terra cotta             91       4      15     101
Muted mustard yellow    84       8      13      87
Aubergine               83       2      13     111
Brick red               88       7      12     110
Black                   84       5       9     124
Ruby Red                90       3       8     107
Mauve                   86       5       7      95
Burnt orange            93       4       5     112
Brown                   97       5       3      89

In [ ]:
# Using the "idmax" method, get the index with the maximum value for each column
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

pivot1.idxmax()

Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object

In [ ]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Item Purchased" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable and display it

pivot2 = pd.pivot_table(data = df, columns = 'Season', index = 'Item Purchased', values= 'Location', aggfunc= len)

In [ ]:
# Using the "idmax" method, get the index with the maximum value for each column on the pivot table containing clothing items & season
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

pivot2.idxmax()

Season
Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings
dtype: object

In [26]:
# Create two data-frames for promo-code users and non-promo-code users using boolean indexing
# Save these two dataframes into seperate variables 
# Documentation: https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing
# Hint: Users that have used a promo code have the value "Yes" in the column "Promo Code Used"

promo = df[df['Promo Code Used'] == "Yes"]
promo

Age  Gender Item Purchased  Purchase Amount (USD)       Location Size  \
1785  36.0  Female  Running Shoes              62.442101  Massachusetts    L   
1786  57.0    Male       Backpack              51.969839       Michigan    M   
1787  26.0    Male          Socks              54.899050      Wisconsin    M   
1788  36.0  Female        Handbag              46.883976       Delaware    S   
1789  34.0  Female       Backpack              45.171082          Texas    L   
...    ...     ...            ...                    ...            ...  ...   
3153  39.0    Male  Running Shoes              55.636680         Kansas    L   
3154  43.0  Female     Sunglasses              61.610602       Colorado    S   
3155  37.0    Male        Pajamas              44.600556         Alaska    S   
3156  39.0    Male         Hoodie              45.343778       Illinois    S   
3157  21.0  Female         Gloves              49.439181   North Dakota    M   

             Color  Season Review Rating   Shipping Type Promo Code Used  \
1785         White  Summer       Present        Standard             Yes   
1786     Aubergine    Fall       Missing  2-Day Shipping             Yes   
1787         Black    Fall       Missing  2-Day Shipping             Yes   
1788         Mauve  Summer       Missing  2-Day Shipping             Yes   
1789         Mauve  Winter       Present  2-Day Shipping             Yes   
...            ...     ...           ...             ...             ...   
3153         White  Summer       Missing  2-Day Shipping             Yes   
3154  Burnt orange    Fall       Missing  2-Day Shipping             Yes   
3155     Aubergine  Winter       Present  2-Day Shipping             Yes   
3156     Brick red    Fall       Missing  2-Day Shipping             Yes   
3157     Aubergine  Winter       Present  2-Day Shipping             Yes   

      Previous Purchases Payment Method  
1785                   0    Credit Card  
1786                   1    Credit Card  
1787                   0    Credit Card  
1788                   2    Credit Card  
1789                  13    Credit Card  
...                  ...            ...  
3153                  11          Venmo  
3154                   7          Venmo  
3155                  20          Venmo  
3156                   3          Venmo  
3157                   2          Venmo  

[1373 rows x 13 columns]

In [24]:
promonot = df[df['Promo Code Used'] == "No"]
promonot

Age  Gender Item Purchased  Purchase Amount (USD)        Location Size  \
0     31.0    Male       Leggings              24.231704          Nevada    M   
1     38.0    Male     Sunglasses              36.545487          Oregon    S   
2     26.0    Male       Leggings              23.595139          Nevada   XL   
3     43.0    Male          Dress              34.079015      California    M   
4     29.0    Male         Shorts              23.796439       Minnesota    M   
...    ...     ...            ...                    ...             ...  ...   
1780  36.0  Female          Socks              32.587494         Arizona   XL   
1781  42.0    Male         Gloves              29.674884         Indiana    L   
1782  52.0    Male        Handbag              33.993693         Arizona    S   
1783  24.0    Male         Hoodie              30.865944        Kentucky    S   
1784  27.0    Male          Boots              29.339384  North Carolina    L   

                     Color  Season Review Rating Shipping Type  \
0              Terra cotta  Winter       Present      Standard   
1                Aubergine  Summer       Missing      Standard   
2                    Brown  Winter       Missing      Standard   
3              Terra cotta    Fall       Present      Standard   
4                 Lavender  Summer       Present       Express   
...                    ...     ...           ...           ...   
1780  Muted mustard yellow  Winter       Missing      Standard   
1781           Terra cotta  Winter       Missing      Standard   
1782             Baby blue  Spring       Missing      Standard   
1783                 Mauve  Winter       Missing      Standard   
1784             Brick red  Winter       Present      Standard   

     Promo Code Used  Previous Purchases Payment Method  
0                 No                   0    Credit Card  
1                 No                   0    Credit Card  
2                 No                   3    Credit Card  
3                 No                   0    Credit Card  
4                 No                   0    Credit Card  
...              ...                 ...            ...  
1780              No                  10          Venmo  
1781              No                   1          Venmo  
1782              No                   0          Venmo  
1783              No                  16          Venmo  
1784              No                   0          Venmo  

[1785 rows x 13 columns]

In [27]:
# Print the shape of the promo code users dataframe. This should be (1373, 13)

promo.shape

(1373, 13)

In [28]:
# Print the average purchase size of promo code users dataframe. This should be 50.06747998236351
promo['Purchase Amount (USD)'].mean()


50.06747998236351

In [29]:
# Print the shape of the non promo code users dataframe. This should be (1785, 13)

promonot.shape

(1785, 13)

In [30]:
# Print the average purchase size of non promo code users dataframe. This should be 30.155750765104113

promonot['Purchase Amount (USD)'].mean()

30.155750765104113

Formulate a null hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

H0: There is no difference in spending amount between promo-code users & non-promo-code users.


Formulate an alternative hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

H1: Promo-code users on average spend more money on purchases compared to non-promo code users. 

In [39]:
# Using the "ttest_ind" method from "scipy.stats", run a T-Test between the dollar amount spent of promo-code users and non promo-code users
# The t-statistic you should observe is 93.2464714350831
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

a = promo['Purchase Amount (USD)']     
b = promonot['Purchase Amount (USD)']
result = stats.ttest_ind(a, b, axis = 0, equal_var= True, alternative = 'two-sided')

print('t-test statistic:', result[0])
print('p-value:', '{:0.10f}'.format(result[1]))


t-test statistic: 93.2464714350831
p-value: 0.0000000000


In [ ]:
# Group your dataframe according to the "Review Rating" column. This should produce two groups for the values "Missing" & "Present"
# Save this grouped dataframe into a new variable named "g_review"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

g_review = df.groupby('Review Rating')
g_review


In [ ]:
# Print out the keys of this dataframe for validation. This should output "dict_keys(['Missing', 'Present'])"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.keys.html
g_review_validation = df.groupby('Review Rating').size().to_dict()
g_review_validation.keys()

dict_keys(['Missing', 'Present'])

In [ ]:
# Calculate the average of the "Purchase Amount (USD)", "Age", and "Previous Purchases" columns in our grouped dataframe
# output the values for display
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
# Documentation: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html

g_review[['Purchase Amount (USD)', 'Age','Previous Purchases']].mean()


Purchase Amount (USD)        Age  Previous Purchases
Review Rating                                                      
Missing                    38.923623  34.080198            5.344059
Present                    38.615947  34.371705            6.778559

In [ ]:
# Create a new column called "Prev Purchase Group" on your original dataframe which seperates your dataframe into 
# two groups of shoppers based on their "Previous Purchases" column using the "pd.cut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.cut.html

df['Prev Purchase Group'] = pd.cut(df['Previous Purchases'], bins = 2, labels = ['Low', 'High'])
df[['Previous Purchases', 'Prev Purchase Group']]



Previous Purchases Prev Purchase Group
0                      0                 Low
1                      0                 Low
2                      3                 Low
3                      0                 Low
4                      0                 Low
...                  ...                 ...
3153                  11                High
3154                   7                 Low
3155                  20                High
3156                   3                 Low
3157                   2                 Low

[3158 rows x 2 columns]

In [ ]:
# Using the "value_counts" function, count how many Missing & Present values are in the "Low" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

df[df['Prev Purchase Group'] == 'Low']['Review Rating'].value_counts()

#Using ->  df.groupby('Prev Purchase Group')['Review Rating'].value_counts()  to double check that correct group values are displayed for "Low"

Review Rating
Missing    1686
Present     840
Name: count, dtype: int64

In [ ]:
# Using the "value_counts" function, count how many Missing & Present values are in the "High" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

df[df['Prev Purchase Group'] == 'High']['Review Rating'].value_counts()


Review Rating
Missing    334
Present    298
Name: count, dtype: int64

## Data Analysis II

In the next section, answer the primary analytical questions in your own words.

### Q1

What are the top three colors for Fall & Winter?

The top three colors for Fall were: Brown, Burnt Orange, and Terra Cotta, while the top three colors for Winter were: Black, Burnt Orange, and Aubergine.

### Q2

What are the top three colors for Spring & Summer?

The top three colors for Spring were Baby Blue, White, and Lavender, while the top three colors for Summer were Lavender, Lemon Yellow, and Baby Blue. 

### Q3

What is the most popular clothing item by season?

Backpacks were the most popular items purchased in the Fall while running shoes were the most popular item purchased in the Spring. Additionally, shorts were the most popular item sold in the Summer, while leggings were the most popular item sold in the Winter. 

### Q4

Observing the p-value that we got when running our t-test between promo-code and non-promo-code users, what can we conclude regarding our null hypothesis?

Based on the p-value (0.0), we can conclude that the differences observed in promo-code and non-promo code users unlikely resulted from random chance. Thus, we reject the null hypothesis. 

### Q5

Observe the `value_counts` output for the "Review Rating" column for both your "High" and "Low" groups. Proportionally speaking, which group is more likely to leave a review? Why might this be happening from the "human"-perspective? Rationalizations are ok at this point, even if they aren't backed up by data.

Proportionlly speaking, people in the "High" Previous Purchase Group are more likely to leave a review compared to individuals in the "Low" Previous Purchase Group. The proportion of people in the "Low" Previous Purchase Group who left a review is 0.3325, while the proportion of people in the "High" Previous Purchase Group is 0.4715. Sampling bias might account for the higher proportion of review engagement displayed within the "High" Previous Purchase group. It is possible that those in the "High" Previous Purchase group are more likely to leave reviews since they are more consistent buyers and possess a level of familiarity to the company's products. However, there might be a higher number of people in the "Low" group leaving reviews (840) compared to the "High group (298), as there might simply be a larger amount of people in population who are either new users to site, or who only purchase items from the site sparingly. 